air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -77,26 +92,81 @@ fn select_by_path_from_stream<'value>(
         }
         ValueAccessor::Error => unreachable!("should not execute if parsing succeeded. QED."),
     };
-
+    let idx = idx as usize;
     let stream_size = stream.len();
     let value = lambda_to_execution_error!(stream
         .peekable()
-        .nth(idx as usize)
+        .nth(idx)
         .ok_or(LambdaError::CanonStreamNotHaveEnoughValues { stream_size, idx }))?;
 
     let result = select_by_path_from_scalar(value, body.iter(), exec_ctx)?;
-    let select_result = StreamSelectResult::from_cow(result, idx);
+    let select_result = LambdaResult::from_cow(result, idx);
     Ok(select_result)
 }
 
+fn select_by_path_from_canon_map<'value>(
+    canon_map: &'value CanonStreamMap<'_>,
+    lambda: &NonEmpty<ValueAccessor<'_>>,
+    exec_ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<Cow<'value, JValue>> {
+    use crate::execution_step::value_types::CanonStream;
+
+    let (prefix, body) = lambda.split_first();
+
+    // HashMap<'map>::get(key: &'key K) forces key's lifetime 'key to be as good as 'map.
+    // This variance-derived requirement forces StreamMapKey<'static> here.
+    // See https://github.com/rust-lang/rust/issues/80389#issuecomment-752067798
+    // for the details.
+    let stream_map_key: StreamMapKey<'_> = match prefix {
+        ValueAccessor::ArrayAccess { idx } => (*idx).into(),
+        ValueAccessor::FieldAccessByName { field_name } => (*field_name).to_owned().into(),
+        ValueAccessor::FieldAccessByScalar { scalar_name } => {
+            let scalar = exec_ctx.scalars.get_value(scalar_name)?;
+            lambda_to_execution_error!(try_scalar_ref_as_stream_map_key(scalar))?
+        }
+        ValueAccessor::Error => unreachable!("should not execute if parsing succeeded. QED."),
+    };
+    let canon_stream_opt = canon_map.index(&stream_map_key);
+    // There will be an empty canon stream if the key was not found.
+    let result = match canon_stream_opt {
```

Comment:
You can avoid this match by constructing empty canon stream with `ok_or_else` on the 129th line.